pkg/uploadproxy/uploadproxy.go

```diff
@@ -221,47 +222,56 @@ func (app *uploadProxyApp) handleUploadRequest(w http.ResponseWriter, r *http.Re
 		return
 	}
 
-	app.proxyUploadRequest(tokenData.Namespace, tokenData.Name, uploadPath, w, r)
+	app.proxyUploadRequest(uploadPath, w, r)
 }
 
-func (app *uploadProxyApp) resolveUploadPath(pvcName, pvcNamespace, defaultPath string) (string, error) {
-	pvc, err := app.client.CoreV1().PersistentVolumeClaims(pvcNamespace).Get(context.TODO(), pvcName, metav1.GetOptions{})
-	if err != nil {
-		if k8serrors.IsNotFound(err) {
-			return "", fmt.Errorf("rejecting Upload Request for PVC %s that doesn't exist", pvcName)
-		}
-
-		return "", err
-	}
-
-	contentType, found := pvc.Annotations[cc.AnnContentType]
-	if !found {
-		return defaultPath, nil
-	}
-
+func (app *uploadProxyApp) resolveUploadPath(pvc *v1.PersistentVolumeClaim, pvcName, defaultPath string) (string, error) {
+	var path string
+	contentType, _ := pvc.Annotations[cc.AnnContentType]
 	switch contentType {
 	case string(cdiv1.DataVolumeKubeVirt), "":
-		return defaultPath, nil
+		path = defaultPath
 	case string(cdiv1.DataVolumeArchive):
 		if strings.Contains(defaultPath, "alpha") {
-			return common.UploadArchiveAlphaPath, nil
+			path = common.UploadArchiveAlphaPath
+		} else {
+			path = common.UploadArchivePath
 		}
-		return common.UploadArchivePath, nil
 	default:
 		return "", fmt.Errorf("rejecting upload request for PVC %s - upload content-type %s is invalid", pvcName, contentType)
 	}
+
+	return app.urlResolver(pvc.Namespace, pvc.Name, path), nil
 }
 
-func (app *uploadProxyApp) uploadReady(pvcName, pvcNamespace string) error {
-	return wait.PollImmediate(waitReadyImterval, waitReadyTime, func() (bool, error) {
-		pvc, err := app.client.CoreV1().PersistentVolumeClaims(pvcNamespace).Get(context.TODO(), pvcName, metav1.GetOptions{})
+func (app *uploadProxyApp) uploadReady(pvcName, pvcNamespace string) (*v1.PersistentVolumeClaim, error) {
+	var pvc *v1.PersistentVolumeClaim
+	err := wait.PollImmediate(waitReadyImterval, waitReadyTime, func() (bool, error) {
+		var err error
+		pvc, err = app.client.CoreV1().PersistentVolumeClaims(pvcNamespace).Get(context.TODO(), pvcName, metav1.GetOptions{})
 		if err != nil {
 			if k8serrors.IsNotFound(err) {
 				return false, fmt.Errorf("rejecting Upload Request for PVC %s that doesn't exist", pvcName)
 			}
 
 			return false, err
 		}
+		// If using upload populator then need to check upload possibility to the PVC'
+		if populators.IsPVCDataSourceRefKind(pvc, cdiv1.VolumeUploadSourceRef) {
+			pvcPrimeName, ok := pvc.Annotations[populators.AnnPVCPrimeName]
```

Comment:
I think this is fine but a little extra.  I don't think it is necessary to check the datasourceref.  Also the "pvc prime" naming betrays populator implementation details.

Reply:
a. you mean just check if the annotation exist? right it works, the way it was before required this check, I can remove.
b. we need to check that pvc prime is ready for upload same as we do for regular pvc so we need to know eventually the pvc prime name here, not sure what is an alternative, putting the service name and inferring from it the pvc prime name? seems redundant to me, I was in favor of using the pvcprimename function like I did before and refer from using all this annotation but..
what do you suggest doing otherwise?